src/main/kotlin/org/rust/lang/core/psi/RsFile.kt

```diff
@@ -98,7 +100,13 @@ class RsFile(
                 val value = doGetCachedData()
                 // Note: if the cached result is invalidated, then the cached result from `memExpansionResult`
                 // must also be invalidated, so keep them in sync
-                CachedValueProvider.Result(value, rustStructureOrAnyPsiModificationTracker)
+                val modificationTracker: Any = when {
+                    /** See [rustStructureOrAnyPsiModificationTracker] */
+                    virtualFile is VirtualFileWindow -> PsiModificationTracker.MODIFICATION_COUNT
+                    value.crate?.origin == PackageOrigin.WORKSPACE -> project.rustStructureModificationTracker
+                    else -> project.rustPsiManager.rustStructureModificationTrackerInDependencies
+                }
```

Comment:
Do I understand correctly that it is optimized version of `rustStructureOrAnyPsiModificationTracker`?

Maybe rewrite it like this:

```rust
val RsElement.rustStructureOrAnyPsiModificationTracker: Any
    get() = rustStructureOrAnyPsiModificationTracker(null)

fun RsElement.rustStructureOrAnyPsiModificationTracker(crate: Crate?): Any {
```

Reply:
It's rather needed to avoid an infinite recursion :D

Reply:
Passing `Crate?` will still lead to an infinite recursion in the case when `crate` is null